compiler/src/dotty/tools/dotc/transform/MacroAnnotations.scala

```diff
@@ -133,14 +154,13 @@ class MacroAnnotations(phase: IdentityDenotTransformer):
       report.error(i"macro annotation $annot added $sym with an inconsistent owner. Expected it to be owned by ${annotated.owner} but was owned by ${sym.owner}.", annot.tree)
     else if annotated.isClass && annotated.owner.is(Package) /*&& !sym.isClass*/ then
       report.error(i"macro annotation can not add top-level ${sym.showKind}. $annot tried to add $sym.", annot.tree)
+  end checkMacroDef
 
-  /**
-   * Enter the symbols generated by MacroAnnotations
-   */
-  private def enterMissingSymbols(tree: DefTree)(using Context) = new TreeTraverser {
+  /** Enter the symbols generated by MacroAnnotations */
+  def enterMissingSymbols(tree: MemberDef, phase: DenotTransformer)(using Context) = new TreeTraverser {
     def traverse(tree: tpd.Tree)(using Context): Unit = tree match
       case tdef @ TypeDef(_, template: Template) =>
-        val isSymbolInDecls = tdef.symbol.asClass.info.decls.toList.toSet
+        val isSymbolInDecls = atNextPhase(tdef.symbol.asClass.info.decls.toList.toSet)
```

Comment:
We avoid double definitions by looking at the already defined symbols in the future. If we don't do that, the following will result in a double definition.
- Process a definition
- Process the same definition as the companion
In both cases, we will call `enterMissingSymbols` resulting in a wrong double definition error